# HW3B - Pandas Fundamentals

See Canvas for details on how to complete and submit this assignment.

## Introduction

This assignment transitions you from NumPy's numerical array operations to Pandas' powerful tabular data manipulation. While NumPy excels at homogeneous numerical arrays, Pandas is designed for the heterogeneous, labeled data that characterizes most real-world datasets—mixing dates, categories, numbers, and text within the same table.

You'll work with real bike share data from Chicago's Divvy system to answer questions about urban transportation patterns. Through three progressively complex problems—exploring usage patterns, analyzing rider behavior, and conducting temporal analysis—you'll discover why Pandas has become the standard tool for data analysis in Python.

The assignment emphasizes Pandas' design philosophy: named column access, explicit indexing methods (loc/iloc), handling missing data, and method chaining for readable data pipelines. You'll also see how Pandas builds on NumPy while adding the structure and convenience needed for practical data science work.

This assignment should take 3-5 hours to complete.

Before submitting, ensure your notebook:

- Runs completely with "Kernel → Restart & Run All"
- Includes thoughtful responses to all interpretation questions
- Uses clear variable names and follows good coding practices
- Shows your work (don't just print final answers)

### Learning Objectives

By completing this assignment, you will be able to:

1. **Construct and manipulate Pandas data structures**
   - Create DataFrames from dictionaries and CSV files
   - Distinguish between Series and DataFrame objects
   - Set and reset index structures appropriately
   - Understand when operations return views vs copies
2. **Apply explicit indexing paradigms**
   - Use `loc[]` for label-based data access
   - Use `iloc[]` for position-based data access
   - Access columns using bracket notation
   - Explain when each indexing method is appropriate
3. **Diagnose and explore datasets systematically**
   - Use `info()`, `describe()`, `head()`, and `dtypes` to understand data structure
   - Identify missing values with `isna()` and `notna()`
   - Calculate summary statistics across different axes
   - Interpret value distributions with `value_counts()`
4. **Filter data with boolean indexing and queries**
   - Combine multiple conditions with `&`, `|`, and `~` operators
   - Use `isin()` for membership testing
   - Apply `query()` for readable complex filters
   - Understand how index alignment affects operations
5. **Work with datetime data**
   - Parse dates during CSV loading
   - Extract temporal components with the `.dt` accessor
   - Filter data by date ranges
   - Create time-based derived features
6. **Connect Pandas patterns to data analysis workflows**
   - Formulate questions that data can answer
   - Choose appropriate methods for different analysis tasks
   - Interpret results in domain context
   - Recognize when vectorized operations outperform apply()

### Generative AI Allowance

You may use GenAI tools for brainstorming, explanations, and code sketches if you disclose it, understand it, and validate it. Your submission must represent your own work and you are solely responsible for its correctness.

### Scoring

Total of 90 points available, will be graded out of 80. Scores of >100% are allowed.

Distribution:

- Tasks: 48 pts
- Interpretation: 32 pts
- Reflection: 10 pts

Points by Problem:

- Problem 1: 3 tasks, 10 pts
- Problem 2: 4 tasks, 14 pts
- Problem 3: 4 tasks, 14 pts
- Problem 4: 3 tasks, 10 pts

Interpretation Questions:

- Problem 1: 3 questions, 8 pts
- Problem 2: 4 questions, 8 pts
- Problem 3: 3 questions, 8 pts
- Problem 4: 3 questions, 8 pts

Graduate differentiation: poor follow-up responses will result in up to a 5pt deduction for that problem.

## Dataset: Chicago Divvy Bike Share

The dataset you will analyze is based on real trip information from Divvy, Chicago's bike share system. It contains individual trips with start/end times, station information, and rider type.

Dataset homepage: https://divvybikes.com/system-data

Each trip includes:

- Trip start and end times (datetime)
- Start and end station names and IDs
- Rider type (member vs casual)
- Bike type (classic, electric, or docked)

Chicago's Department of Transportation uses this data to optimize station placement, understand usage patterns, and improve service. You'll explore similar questions that real transportation analysts investigate.

## Problems

### Problem 1: Creating DataFrames from Scratch

Before loading data from files, you need to understand how Pandas structures are built. In this problem, you'll create Series and DataFrames manually using Python's built-in data structures. This is a quick warmup to establish the fundamentals.

#### Task 1a: Create a Series

Create a Series called `temperatures` representing daily high temperatures for a week:

- Monday: 72°F
- Tuesday: 75°F  
- Wednesday: 68°F
- Thursday: 71°F
- Friday: 73°F

Use the day names as the index. Print the Series and its data type.

##### Your Code

In [20]:
import pandas as pd

temperatures = pd.Series(['72°F', '75°F', '68°F', '71°F', '73°F'], index = ["Monday", "Tuesday", "Wednesday", "Thursday", "Friday"])
print(f'The temperature series is\n{temperatures}')
print(f'The data type for the series is "{temperatures.dtype}"')

The temperature series is
Monday       72°F
Tuesday      75°F
Wednesday    68°F
Thursday     71°F
Friday       73°F
dtype: object
The data type for the series is "object"


#### Task 1b: Create a DataFrame from a Dictionary

Create a DataFrame called `products` with the following data:

| product | price | quantity |
|---------|-------|----------|
| Widget  | 19.99 | 100 |
| Gadget  | 24.99 | 75 |
| Doohickey | 12.49 | 150 |

Use a dictionary where keys are column names and values are lists. Print the DataFrame and report its shape.

##### Your Code

In [15]:
data = { 
           'product': ['Widget', 'Gadget', 'Doohickey'],
           'price': [19.99, 24.99, 12.49],
           'quantity' : [100, 75, 150]
}
products= pd.DataFrame(data)
print(products)
print(f'The shape of the DataFrame is {products.shape}')


     product  price  quantity
0     Widget  19.99       100
1     Gadget  24.99        75
2  Doohickey  12.49       150
The shape of the DataFrame is (3, 3)


#### Task 1c: Access DataFrame Elements

Using the `products` DataFrame from Task 1b, extract and print:

1. The `price` column as a Series
2. The `product` and `quantity` columns as a DataFrame (using a list of column names)

##### Your Code

In [16]:
print('The price column is ')
print(products['price'])
print('\nThe product and quantity columns are ')
print(products[['product', 'quantity']])


The price column is 
0    19.99
1    24.99
2    12.49
Name: price, dtype: float64

The product and quantity columns are 
     product  quantity
0     Widget       100
1     Gadget        75
2  Doohickey       150


#### Interpretation

Answer the following questions (briefly / concisely) in the markdown cell below:

1. Data structure mapping: When you create a DataFrame from a dictionary (like in Task 1b), what do the dictionary keys become? What do the values become?
2. Bracket notation: Why does `df['price']` return a Series, but `df[['price']]` return a DataFrame? What's the difference in what you're asking for?
3. Index purpose: In Task 1a, you used day names as the index instead of default numbers (0, 1, 2...). When would a custom index like this be more useful than the default numeric index?

##### Your Answers

1. When I created a DataFrame from a dictionary, the keys become the column names and the values become the data in those columns. And values must be also list.
2. So, when I write `df['price']`, I am basically asking for a specific column named **price** which becomes series in output. But when I write `df[['price']]`, I am basically asking for a subset of the DataFrame which itself also a DataFrame i.e. 2d output.
3. This custom index is very useful in case of creating database. For example, I am creating a database for 100 people where I want to know their names, age, color tone, race, etc. If I use default numbers for this attributes it is very difficult to decrypt data for particular person. Custom indexing with columns as attribute names are very helpful in this case.

### Problem 2: Loading and Initial Exploration

Before starting this problem, make sure you are working in a copy of this file in the `my_repo` folder you created in HW2a. You must also have a copy of the file `202410-divvy-tripdata-100k.csv` in a subdirectory called `data`. That file structure is illustrated below.

```text
~/insy6500/my_repo
└── homework
    ├── data
    │   └── 202410-divvy-tripdata-100k.csv
    └── hw3b.ipynb
```

#### Task 2a: Load and Understand Raw Data

Start by loading the data "as-is" to get a general understanding of the overall structure and how Pandas interprets it by default.

Note on file paths: The provided code uses `Path` from Python's `pathlib` module to handle file paths. Path objects work consistently across operating systems (Windows uses backslashes `\`, Mac/Linux use forward slashes `/`), automatically using the correct separator for your system. The provided code defines `csv_path` which should be used as the filename in your `pd.read_csv` to load the data file.

1. Use `pd.read_csv` to load `csv_path` (provided below) without specifying any other arguments. Assign it to the variable `df_raw`.
2. Use the methods we described in class to explore the shape, structure, types, etc. of the data. In particular, consider which columns represent dates or categories.
3. Note the amount of memory used by the dataset. See the section on memory diagnostics in notebook 07a for appropriate code snippets using `memory_usage`.

##### Your Code

In [5]:
import pandas as pd
import numpy as np
from pathlib import Path

# create a OS-independent pointer to the csv file created by Setup
csv_path = Path('./data/202410-divvy-tripdata-100k.csv')

# load and explore the data below (create additional code / markdown cells as necessary)
df_raw= pd.read_csv(csv_path)
print('This shows the basic information for the dataframe')
df_raw.info()
print(f"The loaded data is {df_raw.shape} in shape")
print(f"The loaded data has {df_raw.ndim} dimension")
print(f"The data type for the loaded file is\n{df_raw.dtypes}")
print(f"The summary statistics for numeric columns are\n{df_raw.describe()}")
print(f"The first five rows of the loaded dataframe are\n{df_raw.head()}")
print(f"The last five rows of the loaded dataframe are\n{df_raw.tail()}")
# Check memory usage by column
print("Memory usage by column:")
print(df_raw.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{df_raw.memory_usage(deep=True).sum() / 1024**2:.2f} MB")

This shows the basic information for the dataframe
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   ride_id             100000 non-null  object 
 1   rideable_type       100000 non-null  object 
 2   started_at          100000 non-null  object 
 3   ended_at            100000 non-null  object 
 4   start_station_name  89623 non-null   object 
 5   start_station_id    89623 non-null   object 
 6   end_station_name    89485 non-null   object 
 7   end_station_id      89485 non-null   object 
 8   start_lat           100000 non-null  float64
 9   start_lng           100000 non-null  float64
 10  end_lat             99913 non-null   float64
 11  end_lng             99913 non-null   float64
 12  member_casual       100000 non-null  object 
dtypes: float64(4), object(9)
memory usage: 9.9+ MB
The loaded data is (100000, 13) in sh

#### Task 2b: Reload with Proper Data Types

1. Repeat step 2a.1 to reload the data. Use the `dtype` and `parse_dates` arguments to properly assign categorical and date types. Assign the result to the variable name `rides`.
2. After loading, use `rides.info()` to confirm the type changes.
3. Use `memory_usage` to compare the resulting size with that from step 2a.3.

##### Your Code

In [6]:
rides = pd.read_csv(csv_path, dtype = {
    'ride_id': 'category',
    'rideable_type': 'category',
    'start_station_name': 'category',
    'start_station_id': 'category',
    'end_station_name': 'category',
    'end_station_id': 'category',
    'member_casual': 'category'},
    parse_dates =['started_at', 'ended_at'])
print(f'The basic informations of the new loaded file are')
rides.info()
# Check memory usage by column
print("Memory usage by column:")
print(rides.memory_usage(deep=True))

print("\nTotal memory usage:")
print(f"{rides.memory_usage(deep=True).sum() / 1024**2:.2f} MB")


The basic informations of the new loaded file are
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 13 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   ride_id             100000 non-null  category      
 1   rideable_type       100000 non-null  category      
 2   started_at          100000 non-null  datetime64[ns]
 3   ended_at            100000 non-null  datetime64[ns]
 4   start_station_name  89623 non-null   category      
 5   start_station_id    89623 non-null   category      
 6   end_station_name    89485 non-null   category      
 7   end_station_id      89485 non-null   category      
 8   start_lat           100000 non-null  float64       
 9   start_lng           100000 non-null  float64       
 10  end_lat             99913 non-null   float64       
 11  end_lng             99913 non-null   float64       
 12  member_casual       100000 non-null  

#### Task 2c: Explore Structure and Missing Data

Using the `rides` DataFrame from Task 2b:

1. Determine the range of starting dates in the dataframe using the `min` and `max` methods.
2. Count the number of missing values in each column. See the section of the same name in lecture 06b.
3. Convert the Series from step 2 to a DataFrame using `.to_frame(name='count')`, then add a column called 'percentage' that calculates the percentage of missing values for each column.

##### Your Code

In [7]:
# task 2c code here...
minimum_starting_dateframe = rides['started_at'].min()
maximum_starting_dateframe = rides['started_at'].max()
print(f'The range of starting dates in the dataframe is {minimum_starting_dateframe} to {maximum_starting_dateframe}')
print(f'The number of missing values in each column are\n{rides.isna().sum()}')
converted_frame = rides.isna().sum().to_frame(name='count')
print(f'The series after getting converted to dataframe is \n{converted_frame}')
converted_frame['percantage'] = converted_frame['count']*100/len(rides)
print(f'The converted_frame after adding the percantage column becomes\n{converted_frame}')

The range of starting dates in the dataframe is 2024-09-30 23:12:01.622000 to 2024-10-31 23:54:02.851000
The number of missing values in each column are
ride_id                   0
rideable_type             0
started_at                0
ended_at                  0
start_station_name    10377
start_station_id      10377
end_station_name      10515
end_station_id        10515
start_lat                 0
start_lng                 0
end_lat                  87
end_lng                  87
member_casual             0
dtype: int64
The series after getting converted to dataframe is 
                    count
ride_id                 0
rideable_type           0
started_at              0
ended_at                0
start_station_name  10377
start_station_id    10377
end_station_name    10515
end_station_id      10515
start_lat               0
start_lng               0
end_lat                87
end_lng                87
member_casual           0
The converted_frame after adding the percantage column

#### Task 2d: Create Trip Duration Column and Set Index

Before setting the index, create a derived column for trip duration:

1. Calculate trip_duration_min by subtracting `started_at` from `ended_at`, then converting to minutes using `.dt.total_seconds() / 60`
3. Display basic statistics (min, max, mean) for the new column using `.describe()`
4. Show the first few rows with `started_at`, `ended_at`, and `trip_duration_min` to verify the calculation
5. Set `started_at` as the DataFrame's index. Verify the change by printing the index and displaying the first few rows.

##### Your Code

In [8]:
# task 2d code here...
rides['trip_duration_min'] = (rides['ended_at']-rides['started_at']).dt.total_seconds() / 60
print(f'The basic statistics for the new column are\n{rides['trip_duration_min'].describe()}')
print(f'The first five rows of the dataframe is\n{rides[['started_at', 'ended_at', 'trip_duration_min']].head()}')
rides = rides.set_index('started_at')
print(f'The index of the dataframe is\n{rides.index}')
print(f'The first five rows of the dataframe is\n{rides.head()}')

The basic statistics for the new column are
count    100000.000000
mean         16.144576
std          52.922539
min           0.006533
25%           5.489271
50%           9.423592
75%          16.407171
max        1499.949717
Name: trip_duration_min, dtype: float64
The first five rows of the dataframe is
               started_at                ended_at  trip_duration_min
0 2024-09-30 23:12:01.622 2024-10-01 00:20:00.674          67.984200
1 2024-09-30 23:19:25.409 2024-10-01 00:42:09.933          82.742067
2 2024-09-30 23:32:24.672 2024-10-01 00:23:18.647          50.899583
3 2024-09-30 23:42:11.207 2024-10-01 00:10:16.831          28.093733
4 2024-09-30 23:49:25.380 2024-10-01 00:06:27.476          17.034933
The index of the dataframe is
DatetimeIndex(['2024-09-30 23:12:01.622000', '2024-09-30 23:19:25.409000',
               '2024-09-30 23:32:24.672000', '2024-09-30 23:42:11.207000',
               '2024-09-30 23:49:25.380000', '2024-09-30 23:49:40.016000',
               '2024-10

#### Interpretation

Reflect on problem 2 and answer (briefly / concisely) the following questions:

1. What types did Pandas assign to `started_at` and `member_casual` in Task 2a? Why might these defaults be problematic?
2. Look at the values in the station ID fields. Based on what you learned about git commit IDs in HW3a, how do you think the station IDs were derived?
3. Explain in your own words what method chaining is, what `df.isna().sum()` does and how it works.
4. Assume you found ~10% missing values in station columns but ~0% in coordinates. What might explain this? How might you handle the affected rows?

##### Your Answers

1. Pandas assigned `object` type for both `started_at` and `member_casual` in Task 2a. However, `started_at` is basically a datetime data. Since it is assigned as `object`, it can not perform datetime operations. For example anything related to date and time, it is not possible to do without changing the type. Again, for `member_casual` it has only two type either `member` or `casual`. So, keeping it as object type data is just waste of memory.
2. The station ID fields look like the git commit IDs. These look like the cryptographic hashes. These are not any sequential numbers. These are more like unique identifiers to identify the unique station with cryptic hashes instead of disclosing the full details of that particular station.
3. Here df is a DataFrame. And isna() function basically finds the NAN values present in each column of the data frame and returns a boolean array basically DataFrame containing the True/False value in each column. And sum() function calculates the total number of NAN values present in each column.
4.  Every place in this earth will have its own coordinate. That's why there is 0% missing value. However, it is possible that the rider started the journey outside the station or ended the journey outside station or both. So, it can have missing values. If station field is a critical attribute for my analysis, I have to drop that row to not to corrupt the data. But if coordinates values are enough for data analysis or machine learning result, I can neglect the missing values in station fields and get done with the coordinates result since it will be there always.

#### Follow-Up (Graduate Students Only)

Compare memory usage results in 2a.3 and 2b.3. What caused the change? Why are these numbers different from what is reported at the bottom of `df.info()`? Which should you use if data size is a concern?

Working with DataFrames typically requires 5-10x the dataset size in available RAM. On a system with 16GB, assuming about 30% overhead from the operating system and other programs, what range of dataset sizes would be safely manageable? Calculate using both 5x (optimistic) and 10x (conservative) multipliers, then explain which you'd recommend for reliable work.

##### Your Answers

In 2a.3 the total memory usage is 57.16MB and in 2b.3 the total memory usage is 14.46 MB. This is mainly because of the data type changing. Normally pandas evaluate category type data as integer codes and this ultimately reduces the repeated calculation of strings. So it reduces the memory usage. Again, datetime64[ns] is also considered as 64 bits integers. So, it is also memory efficient than the object type data.
So, for df.info() it gives the basic information about the dataframe which ultimately is an approximation. So it is almost rounded result. On the other hand, memormy_usage(deep=true) this means this offer deep and accurate result. Since, both of them work differently from each other, the numbers are different. If datasize is a concern, it is better to use memory_usage(deep=true) since it is more accurate and we can change the object  type to category and datetime like I did in 2b.

Since already 30% of the RAM is consumed by the operating system itself, the available RAM is 16*0.7 = 11.2 GB. If I consider 5x(optimistic) datasize then the maximum dataset size is 11.2/5 = 2.24 GB. If I consider 10x (Conservative) datasize then the maximum dataset size is 11.2/10 = 1.12GB. It is always better to be conservative then optimistic. If the system is good it can work on both in optimistic and conservative manner but if the system is itself not that good then it will slow down or stop operating when 5x dataset manner is applied. So I will recommend the maximum dataset size to be 1.12GB.

### Problem 3: Filtering and Transformation

With clean data loaded, you can now filter and transform it to answer specific questions. This problem focuses on Pandas' powerful indexing and filtering capabilities, along with creating derived columns that enable deeper analysis.

You'll continue working with the `rides` DataFrame from Problem 2, which has `started_at` set as the index.

#### Task 3a: Boolean Indexing and Membership Testing

Use boolean indexing and the `isin()` method to answer these questions:

1. How many trips were taken by *members* using *electric bikes*? Use `&` to combine conditions.
2. What percentage of all trips does this represent?
3. How many trips started at any of these three stations: "Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", or "Kingsbury St & Kinzie St"? Use `isin()`.

Note: Remember to use parentheses around each condition when combining with `&`.

##### Your Code

In [9]:
# Task 3a code here...
member_with_electric = rides[(rides['rideable_type'].isin(['electric_bike'])) & (rides['member_casual'].isin(['member']))]
trip_number = len(member_with_electric)
percantage = trip_number*100/len(rides)
three_station = rides[rides['start_station_name'].isin(["Streeter Dr & Grand Ave", "DuSable Lake Shore Dr & Monroe St", "Kingsbury St & Kinzie St"])]
trip_three_station = len(three_station)
print(f"Number of trips taken by members using electric bikes are {trip_number}")
print(f"Percantage of trips taken by members using electric bikes is {percantage}")
print(f"Trip started at the three stations is {trip_three_station}")

Number of trips taken by members using electric bikes are 33121
Percantage of trips taken by members using electric bikes is 33.121
Trip started at the three stations is 2702


#### Task 3b: Create Derived Columns from Datetime

Add two categorical columns to the rides DataFrame based on trip start time:

1. `is_weekend`: Boolean column that is True for Saturday/Sunday trips. Use .dt.dayofweek on the index (Monday=0, Sunday=6).
2. `time_of_day`: String categories based on start hour:
   - "Morning Rush" if hour is 7, 8, or 9
   - "Evening Rush" if hour is 16, 17, or 18
   - "Midday" for all other hours

For step 2, initialize the column to "Midday", then use .loc[mask, 'time_of_day'] with boolean masks to assign rush hour categories. Extract hour using .dt.hour on the index.

After creating both columns, use value_counts() on time_of_day to show the distribution.

##### Your Code

In [10]:
# Task 3b code here...
rides['is_weekend'] = rides.index.dayofweek >= 5
rides['time_of_day'] = 'Midday'
dayhour = rides.index.hour
rides.loc[dayhour.isin([7,8,9]), 'time_of_day'] = 'Morning Rush'
rides.loc[dayhour.isin([16,17,18]), 'time_of_day'] = 'Evening Rush'
print(f"The following is the distribution \n{rides['time_of_day'].value_counts()}")

The following is the distribution 
time_of_day
Midday          55912
Evening Rush    28218
Morning Rush    15870
Name: count, dtype: int64


#### Task 3c: Complex Filtering with query()

Use the `query()` method to find trips that meet **all** of these criteria:
- Casual riders (not members)
- Weekend trips  
- Duration greater than 20 minutes
- Electric bikes

Report:
1. How many trips match these criteria?
2. What percentage of all trips do they represent?
3. What is the average duration of these trips?

Hint: Column names work directly in `query()` strings. Combine conditions with `and`.

##### Your Code

In [11]:
all_criteria_match = rides.query(
    "member_casual == 'casual' and is_weekend == True and trip_duration_min > 20 and rideable_type == 'electric_bike'"
)
no_of_trips = len(all_criteria_match)

percentage_of_trips = (no_of_trips / len(rides)) * 100
avg_duration = all_criteria_match['trip_duration_min'].mean()
print(f"Number of trips matching criteria is {no_of_trips}")
print(f"Percentage of all trips is {percentage_of_trips}%")
print(f"Average trip duration is {avg_duration} minutes")


Number of trips matching criteria is 1501
Percentage of all trips is 1.5010000000000001%
Average trip duration is 40.37352101932046 minutes


#### Task 3d: Explicit Indexing Practice

Practice using `loc[]` and `iloc[]` for different selection tasks:

1. Use `iloc[]` to select the first 10 trips, showing only `member_casual`, `rideable_type`, and `trip_duration_min` columns
2. Use `loc[]` to select trips from October 15-17 (use date strings `'2024-10-15':'2024-10-17'`), showing the same three columns
3. Count how many trips occurred during this date range

Note: When using `iloc[]`, remember it's position-based (0-indexed). When using `loc[]` with the datetime index, you can slice using date strings.

##### Your Code

In [12]:
# Task 3d code here...
first_10_trips = rides.iloc[0:10, [11, 1, 12]]
trips_in_oct = rides.loc['2024-10-15':'2024-10-17', ['member_casual', 'rideable_type', 'trip_duration_min']]
trip_count = len(trips_in_oct)
print(f"The first 10 trips are\n{first_10_trips}")
print(f"Trips from October 15 -17 are\n{trips_in_oct}")
print(f"Number of trips in these date range is {trip_count}")

The first 10 trips are
                        member_casual  rideable_type  trip_duration_min
started_at                                                             
2024-09-30 23:12:01.622        casual  electric_bike          67.984200
2024-09-30 23:19:25.409        casual  electric_bike          82.742067
2024-09-30 23:32:24.672        member   classic_bike          50.899583
2024-09-30 23:42:11.207        casual   classic_bike          28.093733
2024-09-30 23:49:25.380        member  electric_bike          17.034933
2024-09-30 23:49:40.016        member  electric_bike          13.009367
2024-10-01 00:00:53.414        member   classic_bike           2.598817
2024-10-01 00:05:44.954        member  electric_bike           0.013433
2024-10-01 00:06:12.035        member  electric_bike          10.472933
2024-10-01 00:10:03.646        member   classic_bike           7.825683
Trips from October 15 -17 are
                        member_casual  rideable_type  trip_duration_min
started_at 

#### Interpretation

Reflect on this problem and answer (briefly / concisely) the following questions:

1. `isin()` advantages: Compare using `isin(['A', 'B', 'C'])` versus `(col == 'A') | (col == 'B') | (col == 'C')`. Beyond readability, what practical advantage does `isin()` provide when filtering for many values (e.g., 20+ stations)?
2. Conditional assignment order: In Task 3b, why did we initialize all values to "Midday" before assigning rush hour categories? What would go wrong if you assigned categories in a different order, or didn't set a default?
3. `query()` vs boolean indexing: The `query()` method in Task 3c could have been written with boolean indexing instead. When would you choose `query()` over boolean indexing? When might boolean indexing be preferable despite being more verbose?

##### Your Answers

1. `isin()` is more efficient than manual comparison. It can execute faster. Even with large datasets like 20+ stations, the code remains compact and it makes the code run faster. So, large dataset is not a problem at all. But think about using large dataset and manual comparison operation like `&` or `|` or something else. It will take too much time to write the code, the system will take longer duration to execute and even for another programmer it is clumsy to interpret the code.
2. We initialized the code with "Midday" to give every row of the datasheet a default value. If we set rush hour before and then assigned "Midday" it was possible that even the rush hour rows get overwritten and becomes "Midday" which is wrong. So, assigning the hour with order is very important. We can update the rush hour later by giving the conditions which we did. But doing the reverse would provide wrong output.
3. It is easier to write with `query()` when the logic is not complex or does not have too many varibles. It is readable and concise and faster as well. However, if we have some complex logic or some nested conditions where too many variables have to be considered for answering the questions or finding the specific  data, it is better to go with boolean indexing.


#### Follow-Up (Graduate Students Only)

Pandas supports a variety of indexing paradigms, including bracket notation (`df['col']`), label-based indexing (`loc[]`), and position-based indexing (`iloc[]`). The lecture recommended using bracket notation only for columns, and loc/iloc for everything else. Explain the rationale: why is this approach better than using bracket notation for everything, even though `df[0:5]` technically works for row slicing?

##### Your Answers

The differentiation is based on clarity and cleanliness. Though `df[0:5]` does row slicing, it is very much possible that it will behave weird or unnatarual when the index is not numeric. Like in question 2, we set index which is basically `datetime` attribute. The row slicing with `df[0:5]` may not function well. Again, bracket notation is very good in a sense that you just specify the column you want, it will give you all the data in that column. Using `loc[]` or `iloc[]` you don't have to worry about the index, is it numeric or something else, you get the specific row or column data you specify.

### Problem 4: Temporal Analysis and Export

Time-based patterns are crucial for understanding bike share usage. In this problem, you'll analyze when trips occur, how usage differs between rider types, and export filtered results. You'll use the datetime index you set in Problem 2 and the derived columns from Problems 2-3.

#### Task 4a: Identify Temporal Patterns

Use the datetime index to extract temporal components and identify usage patterns:

1. Extract the *hour* from the index and use `value_counts()` to find the most popular hour for trips. Report the peak hour and how many trips occurred during that hour.
2. Extract the *day name* from the index and use `value_counts()` to find the busiest day of the week. Report the day and number of trips.
3. Sort the results from step 2 to show days in order from Monday to Sunday (not by trip count). Use `sort_index()`.

Hint: Use `.dt.hour` and `.dt.day_name()` on the datetime index.

##### Your Code

In [ ]:
# Task 4a code here...


#### Task 4b: Compare Groups with groupby()

Use `groupby()` (introduced in 07a) to compare trip characteristics across different groups:

1. Calculate the average trip duration by rider type (`member_casual`). Which group takes longer trips on average?
2. Calculate the average trip duration by bike type (`rideable_type`). Which bike type has the longest average trip?
3. Count the number of trips by rider type using `groupby()` with `.size()`. Compare this with using `value_counts()` on the `member_casual` column - do they give the same result?

Note: Use single-key groupby only (one column at a time).

##### Your Code

In [ ]:
# Task 4b code here...


#### Task 4c: Filter, Sample, and Export

Create a filtered dataset for weekend electric bike trips and export it:

The provided code once again uses Path to create an `output` directory and constructs the full file path as `output/weekend_electric_trips.csv`. Use the `output_file` variable when calling `.to_csv()`.

1. Filter for trips where `is_weekend == True` and `rideable_type == 'electric_bike'`
2. Use `iloc[]` to select the first 1000 trips from this filtered dataset
3. Use `reset_index()` to convert the datetime index back to a column (so it's included in the export)
4. Export to CSV with filename `weekend_electric_trips.csv`, including only these columns: `started_at`, `ended_at`, `member_casual`, `trip_duration_min`, `time_of_day`
5. Use `index=False` to avoid writing the default numeric index to the file

After exporting, report how many total weekend electric bike trips existed before sampling to 1000.

##### Your Code

In [ ]:
# do not modify this setup code
from pathlib import Path

output_dir = Path('output')
output_dir.mkdir(exist_ok=True)
output_file = output_dir / 'weekend_electric_trips.csv'

# Task 4c code here...
# use the variable `output_file` as the filename for step 4


#### Interpretation

Reflect on this problem and answer the following questions:

1. `groupby() conceptual model`: Explain in your own words what `groupby()` does. Use the phrase "split-apply-combine" in your explanation and describe what happens at each stage.
2. `value_counts()` vs `groupby()`: In Task 4b.3, you compared two approaches for counting trips by rider type. When would you use `value_counts()` versus `groupby().size()`? Is there a situation where only one of them would work?
3. Index management for export: In Task 4c, why did we use `reset_index()` before exporting? What would happen if you exported with the datetime index still in place and used `index=False`?

##### Your Answers

*Problem 4 interpretation here*

#### Follow-Up (Graduate Students Only)

Compare `CSV` and _pickle_ formats for data storage and retrieval.

Pickle is Python's built-in serialization format that saves Python objects exactly as they exist in memory, preserving all data types, structures, and metadata. Unlike CSV (which converts everything to text), pickle is binary (not human readable) and maintains the complete state of your DataFrame. Also, pickle files only work in Python, while CSV is universal. Read more in the [Pandas documentation](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_pickle.html).

The code below investigates an interesting pattern: Do riders take longer trips from scenic lakefront stations even during rush hours? This could indicate tourists or recreational riders using these popular locations for leisure trips during typical commute times. The analysis filters for trips over 15 minutes that started from lakefront stations during morning (7-9am) or evening (4-6pm) rush hours, sorted by duration to see the longest trips first.

Run the code below, then answer the interpretation questions:

In [ ]:
import os

# the following lines were commented out since they were run in 4c
# from pathlib import Path
# output_dir = Path('output')

csv_file = output_dir / 'lakefront_rush_trips.csv'
pickle_file = output_dir / 'lakefront_rush_trips.pkl'

# Filter for interesting pattern: Long trips (>15 min) during rush hours 
# from lakefront stations, sorted by duration
lakefront_rush = (rides
    .loc[(rides.index.hour.isin([7, 8, 9, 16, 17, 18]))]
    .loc[(rides['start_station_name'].str.contains('Lake Shore|Lakefront', 
                                                    case=False, 
                                                    na=False))]
    .loc[rides['trip_duration_min'] > 15]
    .sort_values('trip_duration_min', ascending=False)
    .head(1000)
    .reset_index()
    [['started_at', 'ended_at', 'start_station_name', 'end_station_name',
      'member_casual', 'rideable_type', 'trip_duration_min']]
)

print(f"Found {len(lakefront_rush)} long rush-hour trips from lakefront stations")

# Export to both formats
lakefront_rush.to_csv(csv_file, index=False)
lakefront_rush.to_pickle(pickle_file)

# Compare file sizes
csv_size = os.path.getsize(csv_file) / 1024  # Convert to KB
pickle_size = os.path.getsize(pickle_file) / 1024
print(f"\nCSV file size: {csv_size:.2f} KB")
print(f"Pickle file size: {pickle_size:.2f} KB")
print(f"Size difference: {abs(csv_size - pickle_size):.2f} KB")

# Compare load times
print("\nLoad time comparison:")
print("CSV:")
%timeit pd.read_csv(csv_file)
print("\nPickle:")
%timeit pd.read_pickle(pickle_file)

# Check data type preservation
# Note: CSV load without parse_dates loses datetime types
csv_loaded = pd.read_csv(csv_file)
pickle_loaded = pd.read_pickle(pickle_file)

print("\nData types from CSV (without parse_dates):")
print(csv_loaded.dtypes)
print("\nData types from Pickle:")
print(pickle_loaded.dtypes)

After running the code, answer these questions:

1. Method chaining: The analysis uses method chaining with a specific formatting pattern:

   ```python
   result = (df
       .method1()
       .method2()
       .method3()
   )
   ```

   This wraps the entire chain in parentheses, allowing each method to appear on its own line without backslashes. Discuss why this makes formatting more readable, how it makes debugging easier, how it relates to seeing changes in the code with git diff, and what downsides heavy chaining might have.
3. Data types: Compare the dtypes from CSV versus pickle. What types were preserved by pickle that were lost in CSV? Why is this preservation significant for subsequent analysis?
4. Trade-offs: Given your observations about size, speed, and type preservation, when would you choose pickle over CSV for your work? When would CSV still be the better choice despite pickle's advantages?


*Graduate follow-up interpretation here*

## Reflection

Address the following questions in a markdown cell:

1. NumPy vs Pandas
   - What was the biggest conceptual shift moving from NumPy arrays to Pandas DataFrames?
   - Which Pandas concept was most challenging: indexing (loc/iloc), missing data, datetime operations, or method chaining? How did you work through it?
2. Real Data Experience
   - How did working with real CSV data (with missing values, datetime strings, etc.) differ from hw2b's synthetic NumPy arrays?
   - Based on this assignment, what makes Pandas well-suited for data analysis compared to pure NumPy?
3. Learning & Application
   - Which new skill from this assignment will be most useful for your own data work?
   - On a scale of 1-10, how prepared do you feel to use Pandas for your own projects? What would increase that score?
4. Feedback
   - Time spent: ___ hours (breakdown optional)
   - Most helpful part of the assignment: ___
   - One specific improvement suggestion: ___

### Your Answers

*Reflection here*